tapir/coop/services/MembershipResignationService.py

```diff
@@ -1,53 +1,63 @@
 import datetime
 
 from django.db import transaction
+
 from tapir.accounts.models import TapirUser
 from tapir.coop.models import MembershipResignation, ShareOwnership
 from tapir.shifts.models import (
     ShiftAttendanceTemplate,
     ShiftAttendance,
-    ShiftAccountEntry,
 )
-from tapir.shifts.services.frozen_status_service import FrozenStatusService
-from tapir.shifts.models import ShiftUserData, ShiftAttendanceMode
-
 from tapir.utils.shortcuts import get_timezone_aware_datetime
 
 
 class MembershipResignationService:
     @staticmethod
     @transaction.atomic
-    def update_shifts_and_shares(self, resignation: MembershipResignation):
+    def update_shifts_and_shares(resignation: MembershipResignation):
         tapir_user: TapirUser = getattr(resignation.share_owner, "user", None)
 
         shares = ShareOwnership.objects.filter(share_owner=resignation.share_owner)
-        if not MembershipResignationService.is_resignation_happening_right_away(
-            self, resignation=resignation
-        ):
-            new_end_date = resignation.cancellation_date.replace(day=31, month=12)
-            new_end_date = new_end_date.replace(year=new_end_date.year + 3)
-            shares.update(end_date=new_end_date)
-            return
 
-        MembershipResignationService.resignation_happens_right_away(
-            self, shares=shares, resignation=resignation
-        )
+        match resignation.resignation_type:
+            case MembershipResignation.ResignationType.BUY_BACK:
+                new_end_date = resignation.cancellation_date.replace(day=31, month=12)
+                new_end_date = new_end_date.replace(year=new_end_date.year + 3)
+                shares.update(end_date=new_end_date)
+                return
+            case MembershipResignation.ResignationType.GIFT_TO_COOP:
+                shares.update(end_date=resignation.cancellation_date)
+            case MembershipResignation.ResignationType.TRANSFER:
+                shares_to_create = [
+                    ShareOwnership(
+                        share_owner=resignation.transferring_shares_to,
+                        start_date=resignation.cancellation_date,
+                    )
+                    for _ in shares
+                ]
```

Comment:
What is this for a syntax? Python one? Where the for loop comes after the list????

Reply:
That's [list comprehension](https://docs.python.org/3.12/tutorial/datastructures.html#list-comprehensions)